src/control/lib/spdk/nvme.go

```diff
@@ -105,74 +110,119 @@ type DeviceHealth struct {
 	VolatileWarn    bool
 }
 
-// Discover calls C.nvme_discover which returns
-// pointers to single linked list of ctrlr_t, ns_t and
-// dev_health_t structs.
-// These are converted to slices of Controller, Namespace
-// and DeviceHealth structs.
-func (n *Nvme) Discover() ([]Controller, error) {
-	failLocation := "NVMe Discover(): C.nvme_discover"
+type remFunc func(name string) error
+
+func realRemove(name string) error {
+	return os.Remove(name)
+}
+
+func cleanLockfiles(log logging.Logger, remove remFunc, pciAddrs ...string) error {
+	log.Debugf("removing lockfiles: %v", pciAddrs)
+
+	for _, pciAddr := range pciAddrs {
+		fName := lockfilePathPrefix + pciAddr
 
-	if retPtr := C.nvme_discover(); retPtr != nil {
-		return processReturn(retPtr, failLocation)
+		if err := remove(fName); err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+			return errors.Wrapf(err, "remove %s", fName)
+		}
+
+		log.Debugf("%s removed", fName)
 	}
 
-	return nil, errors.Errorf("%s unexpectedly returned NULL", failLocation)
+	return nil
+}
+
+// wrapCleanErr encapsulates inErr inside any cleanErr.
+func wrapCleanErr(inErr error, cleanErr error) (outErr error) {
+	outErr = inErr
+
+	if cleanErr != nil {
+		outErr = errors.Wrap(inErr, cleanErr.Error())
+		if outErr == nil {
+			outErr = cleanErr
+		}
+	}
+
+	return
+}
+
+// CleanLockfiles removes SPDK lockfiles after binding operations.
+func (n *Nvme) CleanLockfiles(log logging.Logger, pciAddrs ...string) error {
+	return cleanLockfiles(log, realRemove, pciAddrs...)
+}
+
+// Discover NVMe devices accessible by SPDK on a given host.
+//
+// Calls C.nvme_discover which returns pointers to single linked list of
+// ctrlr_t structs. These are converted and returned as Controller slices
+// containing any Namespace and DeviceHealth structs. Afterwards remove
+// lockfile for each discovered device.
+func (n *Nvme) Discover(log logging.Logger) (ctrlrs []Controller, err error) {
+	ctrlrPciAddrs := make([]string, 0, len(ctrlrs))
+
+	ctrlrs, err = processReturn(C.nvme_discover(),
+		"NVMe Discover(): C.nvme_discover")
+	if err != nil {
+		goto clean
+	}
+
+	for _, c := range ctrlrs {
+		ctrlrPciAddrs = append(ctrlrPciAddrs, c.PCIAddr)
+	}
+	log.Debugf("discovered nvme ssds: %v", ctrlrPciAddrs)
+
+clean:
+	err = wrapCleanErr(err, n.CleanLockfiles(log, ctrlrPciAddrs...))
```

Comment:
Might be better to use a helper method to wrap this call, and just return the error from that, instead of using gotos.

Reply:
"might be better"? is there a problem with using gotos? commonly used in the standard library and we use it in our C code.